### PR TITLE
Disallow invalid seed string values

### PR DESCRIPTION
**Pull request by stancld**
## What does this PR do?

See title.
Fixes mypy typing in `utilities.seed.py`

Part of #7037. This PR disentangles large #7218
 
### Does your PR introduce any breaking changes? If yes, please list them.

Passing invalid `seed` values to `seed_everything` is now disallowed.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/seed.py [hunk 1]
```diff
@@ -48,13 +48,13 @@ def seed_everything(seed: Optional[int] = None, workers: bool = False) -> int:
...
-        if seed is None:
-            seed = os.environ.get("PL_GLOBAL_SEED")
-        seed = int(seed)
-    except (TypeError, ValueError):
-        seed = _select_seed_randomly(min_seed_value, max_seed_value)
-        rank_zero_warn(f"No correct seed found, seed set to {seed}")
+    if seed is None:
+        global_seed = os.environ.get("PL_GLOBAL_SEED")
+        if isinstance(global_seed, str) and all(char.isdigit() for char in global_seed):
+            seed = int(global_seed)
```
Comment (awaelchli) [inline pytorch_lightning/utilities/seed.py:54]: I believe the changes here are not equivalent to the previous code. 
If this condition does not hold, we select no seed while previously we would handle it in the except block. 
Anything standing against keeping it? 

Generally I would suggest not letting mypy force changes of our code flow.

Comment (stancld) [reply pytorch_lightning/utilities/seed.py:54]: I've also been suspecting the code behaviour might be a bit changed :/ . The code can be adjusted to ensure the same functionality, however, it's gonna be more complex than the original solution. Maybe mypy can be asked not to check the `seed = int(seed)` statement?

**Comment by tchaton** [comment]
Dear @stancld,

Any updates on this PR ?

Best,
T.C

**Comment by stancld** [comment]
Dear @tchaton:  @awaelchli stressed the importance to keep the functionality the same. In this case, however, satisfying the requirements imposed by mypy leads to much more complex code. Since here, the possible type errors are handled by `try - except` statement, would it be possible just to ask mypy not to check the typing for the line stating `seed = int(seed)`?

**Comment by awaelchli** [comment]
Yes I think we should favor reproducibility here and not the typing. Let's not change the implementation unless there is a bug. In this case ignoring a line there for mypy is well justified imo

**Comment by tchaton** [comment]
Dear @awaelchli  @stancld,

First of all, thanks @stancld for investing time and effort on this PR. This is well appreciated !
What should we do ? Close this PR or add ignore flags for mypy there with a comment on why we made this decision ?

Best.
T.C

**Comment by stancld** [comment]
@tchaton If it's okay I would re-write this PR tomorrow and add two ignoring comments with a supplementary one why this solution is appropriate :)

**Comment by tchaton** [comment]
Hey @stancld,

Thanks for helping out !

Best,
T.C

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/seed.py [hunk 2]
```diff
@@ -47,9 +47,11 @@ def seed_everything(seed: Optional[int] = None, workers: bool = False) -> int:
     min_seed_value = np.iinfo(np.uint32).min
 
     try:
+        # Mypy typing is ignored below as the code simplicity is prefered to mypy correctness. Also, possible errors
```
Comment (carmocca) [inline pytorch_lightning/utilities/seed.py:50]:
```suggestion
        # Mypy typing is ignored below as the code simplicity is preferred to mypy correctness. Also, possible errors
```

File: pytorch_lightning/utilities/seed.py [hunk 3]
```diff
@@ -47,9 +47,11 @@ def seed_everything(seed: Optional[int] = None, workers: bool = False) -> int:
...
+        # are handled by the exception.
         if seed is None:
-            seed = os.environ.get("PL_GLOBAL_SEED")
-        seed = int(seed)
+            seed = os.environ.get("PL_GLOBAL_SEED")  # type: ignore
+        seed = int(seed)  # type: ignore
```
Comment (carmocca) [inline pytorch_lightning/utilities/seed.py:54]: I personally think we could refactor it, the following should match the functionality and comply with mypy

```python
    if seed is None:
        env_seed = os.environ.get("PL_GLOBAL_SEED")
        if env_seed is None:
            seed = _select_seed_randomly(min_seed_value, max_seed_value)
            rank_zero_warn(f"No seed found, seed set to {seed}")
        else:
            try:
                seed = int(env_seed)
            except ValueError:
                seed = _select_seed_randomly(min_seed_value, max_seed_value)
                rank_zero_warn(f"Invalid seed found: {repr(env_seed)}, seed set to {seed}")
    if not isinstance(seed, int):
        seed = int(seed)
```

It limits the scope of the try-except to only the parsing of what's in the environment variable, which is unavoidable as we don't have control over what's written there.

Note that calling `seed_everything("invalid")` would now fail whereas it didn't in the previous version, but I think that case should fail as it violates the typing annotation.

Thoughts @awaelchli ?

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/utilities/seed.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/seed.py:54]: I don't get it fully but make sure that the behavior does not change.

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/seed.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/utilities/seed.py:54]: `seed_everything("invalid")` would change but do you think it should be supported? Passing a non-number string would get you a different random seed each time, so I don't think it should

Comment (awaelchli) [reply pytorch_lightning/utilities/seed.py:54]: yes that's probably safer.

**Pull request merged**